src/org/zaproxy/zap/extension/websocket/WebSocketEventPublisher.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ * 
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ * 
+ * Copyright 2019 The ZAP Development Team
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.websocket;
+
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import org.zaproxy.zap.ZAP;
+import org.zaproxy.zap.eventBus.Event;
+import org.zaproxy.zap.eventBus.EventPublisher;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy.Initiator;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy.State;
+
+/**
+ * A {@link EventPublisher} of websocket events.
+ * @since TODO add version
+ */
+public final class WebSocketEventPublisher implements EventPublisher, WebSocketSenderListener {
+
+    /**
+     * The event sent when a {@code WebSocketMessage} is seen.
+     */
+    public static final String EVENT_MESSAGE = "ws.message";
+    /**
+     * The event sent when a {@code WebSocketProxy} state change occurs.
+     */
+    public static final String EVENT_STATE_CHANGE = "ws.stateChange";
+
+    private static final String FIELD_STATE = "state";
+    private static final String FIELD_CHANNEL_ID = "channelId";
+    private static final String FIELD_LOCAL_SOCKET = "localSocket";
+    private static final String FIELD_REMOTE_SOCKET = "remoteSocket";
+    private static final String FIELD_TIME_IN_MS = "timeSentInMs";
+    private static final String FIELD_OP_CODE = "opCode";
+    private static final String FIELD_OP_CODE_STRING = "opCodeString";
+    private static final String FIELD_DIRECTION = "direction";
+    private static final String FIELD_LENGTH = "length";
+    private static final String FIELD_MSG_ID = "messageId";
+    private static final String FIELD_MSG_SUMMARY = "messageSummary";
+
+    private ExtensionWebSocket extension;
+    private ExecutorService executor;
+
+    public WebSocketEventPublisher (ExtensionWebSocket extension) {
+        this.extension = extension;
+        executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
```

Comment:
- The executor should be shutdown when the extension is unloaded.
 - Could use a lambda for the `ThreadFactory`.